jwst/master_background/master_background_nrs_slits_step.py

```diff
@@ -0,0 +1,235 @@
+            elif self.use_correction_pars:
+                self.log.info('Using pre-calculated correction parameters.')
+                master_background = self.correction_pars['masterbkg_1d']
+                mb_multislit = self.correction_pars['masterbkg_2d']
+            else:
+                self.log.info('Calculating master background')
+                master_background, mb_multislit = self._calc_master_background(data_model)
+
+            # Check that a master background was actually determined.
+            if master_background is None:
+                self.log.info('No master background could be calculated. Skipping.')
+                self.record_step_status(data, 'master_background', False)
+                return data
+
+            # Now apply the de-calibrated background to the original science
+            result = nirspec_utils.apply_master_background(data_model, mb_multislit, inverse=self.inverse)
+
+            # Mark as completed and setup return data
+            self.record_step_status(result, 'master_background', True)
+            self.correction_pars = {
+                'masterbkg_1d': master_background,
+                'masterbkg_2d': mb_multislit
+            }
+            if self.save_background:
+                self.save_model(master_background, suffix='masterbg', force=True)
+                self.save_model(mb_multislit, suffix='masterbg2d', force=True)
+
+        return result
+
+    def set_step_pars(self):
+        """Set substep parameters from the parents substeps"""
+        if not self.parent:
+            return
+
+        steps = ['barshadow', 'flat_field', 'pathloss', 'photom']
+        pars_to_ignore = {
+            'barshadow': ['source_type'],
+            'flat_field': ['save_interpolated_flat'],
+            'pathloss': ['source_type'],
+            'photom': ['source_type']
+        }
+
+        for step in steps:
+            pars = getattr(self.parent, step).get_pars()
+            for par in pars_to_ignore[step] + GLOBAL_PARS_TO_IGNORE:
+                del pars[par]
+            getattr(self, step).update(pars)
+
+    @contextmanager
+    def preserve_pars(self):
+        """Ensure no changes to the Step's parameters"""
+        saved_pars = self.get_pars()
+        try:
+            yield saved_pars
+        finally:
+            self.update(saved_pars)
+
+    def _calc_master_background(self, data, user_background=None):
+        """Calculate master background from background slits
+
+        Parameters
+        ----------
+        data : `~jwst.datamodels.MultiSlitModel`
+            The data to operate on.
+
+        user_background : None, string, or `~jwst.datamodels.CombinedSpecModel`
+            Optional user-supplied master background 1D spectrum, path to file
+            or opened datamodel
+
+        Returns
+        -------
+        masterbkg_1d, masterbkg_2d : `~jwst.datamodels.CombinedSpecModel`, `~jwst.datamodels.MultiSlitModel`
```

Comment:
Is the background just a single file?  Is there a reason to return it in a `MultiSlitModel` when a `SlitModel` would suffice?

Reply:
This is accurate. The `MultiSlitModel` is where all the slits have been replaced with the 2D expanded background for each slit. This is equivalent to the `background_2d` in `MasterBackgroundStep`, but for each slit. See the func `jwst.master_background.nirspec_utils.map_to_science_slits`